### ISSUE TITLE
Incorrect formatting of kilojoules

### ISSUE THREAD

**Issue by fekle** [kipcole9/cldr_units#48]
Hi there,

I'm using this great library to format various units correctly for different languages, and for almost everything I have tested, the unit names are correct.

However, what I've noticed is that there is an issue with the abbreviation for kilojoules. In many languages (and SI, as far as I know), the shorthand is `kJ` with an uppercase `J`, but `cldr_units` always returns `kj`:

```elixir
> Cldr.Unit.to_string!(32, unit: :kilojoule, locale: "de", style: :short)
"32 kj"
> Cldr.Unit.to_string!(32, unit: :kilojoule, locale: "en", style: :short)
"32 kj"
> Cldr.Unit.to_string!(32, unit: :kilojoule, locale: "fr", style: :short)
"32 kj"
> Cldr.Unit.to_string!(32, unit: :kilojoule, locale: nil, style: :short)
"32 kj"
```

As far as I can tell, everything is configured correctly on my end, and the CLDR dataset also specifies "kJ." Is there anything I'm missing here? 

Also, I hope I'm posting this issue in the correct repo; it is not entirely clear to me if the unit name formatting is done in this lib or the base CLDR library. 

Some more information:

```
OS: macos 15.1

elixir:  1.17.3-otp-27
erlang:  27.1.2

ex_cldr: "2.40.1"
ex_cldr_calendars: "1.26.2"
ex_cldr_currencies: "2.16.3"
ex_cldr_dates_times: "2.20.3"
ex_cldr_languages: "0.3.3"
ex_cldr_lists: "2.11.1"
ex_cldr_numbers: "2.33.4"
ex_cldr_plugs: "1.3.3"
ex_cldr_territories: "2.9.0"
ex_cldr_units: "3.17.2"

```

Thank you for this great library, and let me know if there is anything I can do to help or if you need more information!

**Comment by kipcole9** [comment]
Yes, that does look like a bug - and a strange one. I'll see if I can resolve it my day (sun just rising in UTC+11)

**Comment by fekle** [comment]
Thank you!

I found a few more oddities and created a small Minimal Reproduction project here: https://github.com/fekle/ex_cldr_units_repro. I hope this helps!

Specifically, `Cldr.Unit.display_name` does not work for kilojoules, and the formatting of bytes and bits seems odd - see `test/ex_cldr_units_repro_test.exs` in the linked repo.

Greetings from UTC+1 ;)

**Comment by kipcole9** [comment]
Thanks @fekle, I'm still investigating - will do my best to get this resolved by Monday latest.

**Comment by fekle** [comment]
Thanks, and no stress!

**Comment by solars** [comment]
@kipcole9 any news on this? we're also using this lib, I noticed it for 'kN' too

Thanks!

**Comment by kipcole9** [comment]
You've all been very very patient while I work this through. It's now fixed in [this commit](https://github.com/elixir-cldr/cldr_units/commit/a6134cb4b57d5d71b10cb339c85fef7ab7308794). 

You've been very patient - more than reasonably patient - as I work out how to handle this. In the end I took a very pragmatic approach that I'm not super happy with - but it does the required job and isn't too messy.

The solution deliver the result you expect  @fekle for `kilojoule`, but not for you @solars for `kilonewton`. 

The reason is that CLDR has strings for some SI-prefixed units, but only those that are commonly considered an atomic unit. Like `kilometre` and `kilojoule`. 

`kilonewton` doesn't make the cut. So the prefix (`kilo`) and unit (`newton`) are string-ified separately and then combined. It happens that the string for `kilo` in `style: :short` is `k`, not `K`.

The SI-prefixed units for which there is a direct string are:
```elixir
    :kilojoule,
    :kilometer,
    :kilobyte,
    :kilowatt,
    :kilopascal,
    :kilometer_per_hour,
    :kilohertz,
    :kilogram,
    :kilowatt_hour,
    :kilocalorie,
    :kilobit,
    :kilowatt_hour_per_100_kilometer
```

The code depends on updates to `ex_cldr` and `ex_cldr_numbers` so please be patient for 24 hours while I test and publish all the updates.

**Comment by fekle** [comment]
Hi there!

Thank you very much for looking into this, and please, no worries about the time, really :)  
I can only begin to imagine how complex this whole topic is, given the number of units, languages, and the endless list of edge cases.

I tried the updated libraries, and indeed, kilojoule is now correct! 🎉

Furthermore, I updated my small test repo at https://github.com/fekle/ex_cldr_units_repro and noted a few more things that may be of help!

Please let me know if I can be of help.

**Comment by kipcole9** [comment]
@fekle thanks much - very helpful repo. I think I can fix Megoule pretty quickly with the same technique applied for Kilojoule. I'll see if I can know these bugs (for the ones that are bugs - rather than data-related) over the weekend.

**Comment by kipcole9** [comment]
I've published [ex_cldr_units version 3.18.1](https://hex.pm/packages/ex_cldr_units/3.18.1) with the following changelog entry. There is still work to do on some of the examples kindly posted by @fekle. I'm publishing a release now because  also appears to [resolve ](https://github.com/elixir-cldr/cldr_units/commit/019b8bce67fde9088285dac3da6a768b13b33c16) a [compilation issue](https://github.com/elixir-cldr/cldr/issues/256) opened by @TimisRobert.

### Bug Fixes

* Fix compilation failures related to compile-time nested alias issues. Thanks to @TimisRobert for the issues. Closes [ex_cldr #256](https://github.com/elixir-cldr/cldr/issues/256).

* Fix `Cldr.Unit.to_string/2` output for additional SI-prefixed units that have atomic translations. Relates to #48. The additions are:
```elixir
:megabit,
:megabyte,
:megapixel,
:megawatt,
:megapascal,
:megaliter,
:megahertz,
:megabit,
:gigahertz,
:gigabit,
:gigabyte,
:gigawatt,
:terabit,
:terabyte,
:petabyte
```

**Comment by krns** [comment]
Hey, I just updated to 3.19.0 and now a test failes with this output:

```
     left:  "104,86\u00A0GB"
     right: "104,86 GB"
```

After the update there is a `\u00A0`, a non-breaking space between value and unit. The strange thing is, that the other tests for MB, kB etc. are still working with normal spaces. I'm using `to_string!/2` as follows:

```elixir
Cldr.Unit.to_string!(style: :narrow, locale: :de)
```

Is this a bug?

Edit: 

I can reproduce the behaviour like this:

```
iex(5)> Cldr.Unit.new!(:kilobyte, 100) |> Cldr.Unit.to_string!(style: :narrow)
"100 kB"
iex(6)> Cldr.Unit.new!(:gigabyte, 100) |> Cldr.Unit.to_string!(style: :narrow)
"100\u00A0GB
```

**Comment by kipcole9** [comment]
It's not a bug, per se.  The formatting is data-driven and some (in fact only a few) units in the `de` locale have been updated to use a non-breaking-space separator as you've seen.

* Gigabyte and gigabit *do* use an NBSP
* Megabyte, terabyte and petabyte use an ASCII space

Both can be considered correct, but it is jarring to have the inconsistency and I'll ask in the CLDR slack first before opening a ticket on the CLDR repository about that.

The [source data](https://github.com/unicode-org/cldr/blob/main/common/main/de.xml#L10996-L11045) looks like the following and you can see there are inconsistencies even in the `digital-byte` unit where a unit count of `1` is formatted with an ascii space and other pluralisations with an NBSP.

```xml
<unit type="digital-byte">
	<displayName>Byte</displayName>
	<unitPattern count="one">{0} Byte</unitPattern>
	<unitPattern count="other">{0} Byte</unitPattern>
</unit>
<unit type="digital-gigabyte">
	<displayName>Gigabyte</displayName>
	<unitPattern count="one">↑↑↑</unitPattern>
	<unitPattern count="other">{0} GB</unitPattern>
</unit>
<unit type="digital-gigabit">
	<displayName>Gigabit</displayName>
	<unitPattern count="one">↑↑↑</unitPattern>
	<unitPattern count="other">{0} Gb</unitPattern>
</unit>
```

**Comment by krns** [comment]
Oh, thank you very much for the explanation @kipcole9! That is indeed somewhat confusing 😀

**Comment by kipcole9** [comment]
I've noted the inconsistency in the Unicode slack and will update here when (if?)  I hear anything.

**Comment by kipcole9** [comment]
The CLDR team confirmed this should be considered a bug and I have [reported an issue](https://unicode-org.atlassian.net/browse/CLDR-18456) to track this.

**Comment by krns** [comment]
Thanks for the update!

**Comment by kipcole9** [comment]
Closing pending improved data in CLDR 48 coming in September 2025.
